added commands for constraints/worker checking

lethliel opened the pull request:
new osc commands workerinfo and checkconstraints:

`osc workerinfo <workername>`

genreates:

```
<worker hostarch="x86_64" registerserver="http://localhost:5252" workerid="workername">
  <hostlabel>TEST</hostlabel>
  <sandbox>chroot</sandbox>
  <linux>
    <version>4.1.34-33</version>
    <flavor>default</flavor>
  </linux>
  <hardware>
    <cpu>
      <flag>fpu</flag>
      [ ... ]
    </cpu>
    <processors>2</processors>
    <jobs>1</jobs>
  </hardware>
</worker>
```

osc checkconstraints must be called in a package directory. If a repository and a arch is given it returns the names of the workers that can build with the constraints. If no arch and repo is given it sums up the workers for each repo/arch. 

It automatically takes the _constraints file in the directory. The file can also be given as an argument. 
If no file is given or found (or ignored with **--ignore-file**) just the project constraints are used. 

Example output without repo/arch

```
# osc checkconstraints _constraints

Repository                Arch                      Worker
----------                ----                      ------
openSUSE_Leap_42.2        x86_64                    14
openSUSE_Leap_42.1        x86_64                    14
```

Example output with repo and arch

```
# osc checkconstraints openSUSE_Leap_42.1 x86_64 _constraints

Worker
------
x86_64:worker:1
x86_64:worker:2
x86_64:worker:3
[ ... ]
```

@marcus-h: Can you please review? I hope i didn't screw the indentation again. I also used pep8 to check if everything is fine. 
@adrianschroeter: Can you please have a look too?

marcus-h commented:
On 2016-12-29 04:37:39 -0800, Marco Strigl wrote:
> @marcus-h: Can you please review? I hope i didn't screw the indentation again. I also used pep8 to check if everything is fine. 

In general, LGTM! Some comments below.

do_workerinfo can be simplified, if you adjust its signature.
Something like:
```
def do_workerinfo(self, subcmd, opts, worker):
    """your docstr..."""
    apiurl = self.get_api_url()
    print(get_worker_info(apiurl, worker))
```


Also, I would move the output formatting code from check_constraints
into do_checkconstraints. That is, in essence, check_constraints could
be implemented like this:

```
def check_constraints(apiurl, prj, repo, arch, package, constraints_data=None):
    query = {'cmd': 'checkconstraints'}
    query['project'] = prj
    query['package'] = package
    query['repository'] = repo
    query['arch'] = arch
    u = makeurl(apiurl, ['worker'], query)
    f = http_POST(u, data=constraints_data)
    root = ET.fromstring(''.join(f))
    return [node.get('name') for node in root.findall('entry')]
```

Now, the formatting can be done in do_checkconstraints. (The advantage is that
check_constraints can be reused.)

Btw., instead of doing all this xml parsing to get the (repo, arch) pairs of
a project, you can simply use the get_repos_of_project function.
```
for r in get_repos_of_project(apiurl, prj):
    workers = check_constraints(apiurl, prj, r.name, r.arch, package, constraints_data)
    ...
```

Another minor thing: the do_checkconstraints docstr contains some "typos"
("contrains" and "builinfo")

Btw., if there will be more worker related commands in the future, it might
make sense to introduce a "worker" command with several subcommands. For
instance,
- osc worker info foo
- osc worker constraints...

lethliel commented:
On 12/29/2016 11:34 PM, Marcus Hüwe wrote:
> On 2016-12-29 04:37:39 -0800, Marco Strigl wrote:
>> @marcus-h: Can you please review? I hope i didn't screw the indentation again. I also used pep8 to check if everything is fine.
> 
> In general, LGTM! Some comments below.
> 

You are completely right. I have moved the formatting to the do_checkconstraints function and simplified the do_workerinfo function. 
I also use the get_repos_of_project call to get the repo/arch pairs. 

Fixed the typos (shame on me)

> 
> Btw., if there will be more worker related commands in the future, it might
> make sense to introduce a "worker" command with several subcommands. For
> instance,
> - osc worker info foo
> - osc worker constraints...
> 

I think this will be the only two worker related commands in the near future. 

Thank you again for your feedback on my pull requests.

marcus-h commented:
On 2016-12-30 02:06:23 -0800, Marco Strigl wrote:
> On 12/29/2016 11:34 PM, Marcus Hüwe wrote:
> > On 2016-12-29 04:37:39 -0800, Marco Strigl wrote:
> >> @marcus-h: Can you please review? I hope i didn't screw the indentation again. I also used pep8 to check if everything is fine.
> > 
> > In general, LGTM! Some comments below.
> > 
> 
> You are completely right. I have moved the formatting to the do_checkconstraints function and simplified the do_workerinfo function. 
> I also use the get_repos_of_project call to get the repo/arch pairs. 
> 
Looks great! Some last cosmetic remarks for do_checkconstraints (feel
free to ignore them;) )

> +        wd = os.curdir

wd is unused:)

> +        project = package = repository = arch = constraintsfile = None
> +        project = store_read_project('.')
> +        package = store_read_package('.')

Superfluous assignment to project and package.

> +       if not constraintsfile is None:
> +           constraintsfile_data = open(constraintsfile, 'r').read()

"if constraintsfile is not None" is a bit more pythonic:)

> +       else:
> +           if not opts.ignore_file:
> +               if os.path.isfile("_constraints"):
> +                   constraintsfile_data = open("_constraints", 'r').read()
> +               else:
> +                   print("No local _constraints file. Using just the project constraints")

Instead of "else:", you can use "elif not opts.ignore_file: ..." (this
saves a level of indentation)

> +           workers = check_constraints(apiurl, project, repository, arch, package, constraintsfile_data)
> +           for worker in workers:
> +               r.append(worker)

Simply
r = check_constraints(apiurl, project, repository, arch, package, constraintsfile_data)
:)

> > 
> > Btw., if there will be more worker related commands in the future, it might
> > make sense to introduce a "worker" command with several subcommands. For
> > instance,
> > - osc worker info foo
> > - osc worker constraints...
> > 
> 
> I think this will be the only two worker related commands in the near future. 
> 
Ok.

marcus-h commented:
Merged. Thanks a lot!

mmohring commented:
Hi, the corresponding API routes in OBS seem not to be there. Is there a pull request for them ?

lethliel commented:
@mmohring yes. The pull request is https://github.com/openSUSE/open-build-service/pull/2485
